File: src/main/java/com/fouan/actors/view/Actors.java
package com.fouan.actors.view;

import com.fouan.actors.Actor;
import com.fouan.actors.ActorId;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Stream;

public final class Actors {

    private final Map<ActorId, Actor> actors;

    public Actors() {
        this.actors = new HashMap<>();
    }

    public void add(Actor actor) {
        actors.put(actor.getId(), actor);
    }

    public void remove(Actor actor) {
        removeById(actor.getId());
    }

    public void removeById(ActorId actorId) {
        actors.remove(actorId);
    }

    public Optional<Actor> findById(ActorId actorId) {
        return Optional.ofNullable(actors.get(actorId));
    }

    public Stream<Actor> all() {
        return actors.values().stream();
    }


    /*public void update(ActorId actorId, UnaryOperator<Actor> updater) {
        findById(actorId)
            .map(computedActor -> updater.apply(computedActor.actor))
            .ifPresent(actor -> actors.put(actorId, new ComputedActor(actor)));
    }*/
}

File: src/main/java/com/fouan/actors/view/Actors.kt
package com.fouan.actors.view

import com.fouan.actors.Actor
import com.fouan.actors.ActorId
import java.util.function.UnaryOperator

class Actors {
    private val actors: MutableMap<ActorId, Actor> = mutableMapOf()

    fun add(actor: Actor) {
        actors[actor.id] = actor
    }

    fun remove(actor: Actor) {
        removeById(actor.id)
    }

    fun removeById(actorId: ActorId) {
        actors.remove(actorId)
    }

    fun findById(actorId: ActorId): Actor? {
        return actors[actorId]
    }

    fun all(): List<Actor> {
        return actors.values.toList()
    }

    fun update(actorId: ActorId, updater: UnaryOperator<Actor>) {
        val actor = findById(actorId)!!
        actors[actorId] = updater.apply(actor)
    }
}